File: src/main/java/leetcode/topinterview/LeetCode138.java
package leetcode.topinterview;

import common.Node;

import java.util.HashMap;

/**
 * @author wujingxinit@outlook.com
 * @date 2022/9/25 00:01
 */
public class LeetCode138 {

    public Node copyRandomList(Node head) {
        HashMap<Node, Node> map = new HashMap<>();

        Node curr = head;
        while (curr != null) {
            map.put(curr, new Node(curr.val));
            curr = curr.next;
        }

        curr = head;
        while (curr != null) {
            map.get(curr).next = map.get(curr.next);
            map.get(curr).random = map.get(curr.random);
            curr = curr.next;
        }

        return map.get(head);
    }
}

File: go/src/leetcode/topinterview/LeetCode138.go
package topinterview

func copyRandomList(head *Node) *Node {
	if head == nil {
		return nil
	}
	mp := make(map[*Node]*Node)
	var dfs func(node *Node) *Node
	dfs = func(node *Node) *Node {
		if node == nil {
			return nil
		}
		if newNode, ok := mp[node]; ok {
			return newNode
		}
		newNode := &Node{
			Val: node.Val,
		}
		mp[node] = newNode
		newNode.Next = dfs(node.Next)
		newNode.Random = dfs(node.Random)
		return newNode
	}
	return dfs(head)
}
